File: packages/schema/src/user/session.ts
/**
 * Session Schema Definitions
 *
 * IMPORTANT: Follow conventions in utils.ts header comment
 * - Use snake_case for all field names
 * - Use .nullable() for Option<T> fields, not .optional()
 */

import { z } from "zod";
import { idSchema, dateTimeSchema } from "../utils";

export const sessionSchema = z.object({
  id: idSchema,
  user_id: idSchema,
  branch_id: idSchema,
  token: z.string(),
  device_id: z.string().nullable(),
  device_name: z.string().nullable(),
  ip_address: z.string().nullable(),
  user_agent: z.string().nullable(),
  expires_at: dateTimeSchema,
  last_activity_at: dateTimeSchema,
  is_active: z.boolean(),
  created_at: dateTimeSchema,
});

export type Session = z.infer<typeof sessionSchema>;

File: apps/desktop/crates/infrastructure/src/user/entities/session.rs
use pharos_core::{DateTime, Id};
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

/// Session entity for tracking active user sessions
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "sessions")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Id,

    pub user_id: Id,
    pub branch_id: Id,

    #[sea_orm(unique)]
    pub token: String,

    #[sea_orm(nullable)]
    pub device_id: Option<String>,

    #[sea_orm(nullable)]
    pub device_name: Option<String>,

    #[sea_orm(nullable)]
    pub ip_address: Option<String>,

    #[sea_orm(column_type = "Text", nullable)]
    pub user_agent: Option<String>,

    pub expires_at: DateTime,
    pub last_activity_at: DateTime,
    pub is_active: bool,
    pub created_at: DateTime,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::UserId",
        to = "super::user::Column::Id",
        on_delete = "Cascade"
    )]
    User,

    #[sea_orm(
        belongs_to = "crate::branch::entities::branch::Entity",
        from = "Column::BranchId",
        to = "crate::branch::entities::branch::Column::Id"
    )]
    Branch,
}

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    fn new() -> Self {
        Self {
            id: sea_orm::ActiveValue::Set(Id::new()),
            last_activity_at: sea_orm::ActiveValue::Set(DateTime::now()),
            is_active: sea_orm::ActiveValue::Set(true),
            created_at: sea_orm::ActiveValue::Set(DateTime::now()),
            ..Default::default()
        }
    }
}
